File: common/src/main/java/org/sandboxpowered/silica/state/SilicaStateFactory.java
package org.sandboxpowered.silica.state;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSortedMap;
import com.google.common.collect.Lists;
import org.apache.commons.lang3.tuple.Pair;
import org.sandboxpowered.api.state.StateFactory;
import org.sandboxpowered.api.state.property.Property;
import org.sandboxpowered.api.state.property.PropertyContainer;

import java.util.*;
import java.util.function.BiFunction;
import java.util.stream.Stream;

public class SilicaStateFactory<A, B extends PropertyContainer<B>> implements StateFactory<A, B> {
    private final A base;
    private final ImmutableSortedMap<String, Property<?>> propertiesByName;
    private final ImmutableList<B> states;

    public SilicaStateFactory(A base, Map<String, Property<?>> map, BiFunction<A, ImmutableMap<Property<?>, Comparable<?>>, B> function) {
        this.base = base;
        this.propertiesByName = ImmutableSortedMap.copyOf(map);
        Stream<List<Pair<Property<?>, Comparable<?>>>> stream = Stream.of(Collections.emptyList());
        for (Map.Entry<String, Property<?>> entry : propertiesByName.entrySet()) {
            stream = stream.flatMap(list -> entry.getValue().getValues().stream().map(comparable -> {
                List<Pair<Property<?>, Comparable<?>>> list2 = Lists.newArrayList(list);
                list2.add(Pair.of(entry.getValue(), comparable));
                return list2;
            }));
        }

        Map<Map<Property<?>, Comparable<?>>, B> propertyToState = new LinkedHashMap<>();
        List<B> allStates = new ArrayList<>();

        stream.forEach(list -> {
            ImmutableMap<Property<?>, Comparable<?>> propertyValues = list.stream().collect(ImmutableMap.toImmutableMap(Pair::getKey, Pair::getValue));
            B state = function.apply(base, propertyValues);
            propertyToState.put(propertyValues, state);
            allStates.add(state);
        });

        allStates.forEach(state -> {
            if (state instanceof BaseState) {
                ((BaseState<A, B>) state).initTable(propertyToState);
            }
        });
        this.states = ImmutableList.copyOf(allStates);
    }

    @Override
    public B getBaseState() {
        return states.get(0);
    }

    @Override
    public A getBaseObject() {
        return base;
    }

    @Override
    public Collection<B> getValidStates() {
        return Collections.emptySet();
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/state/SilicaStateFactory.kt
package org.sandboxpowered.silica.state

import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.common.collect.ImmutableSortedMap
import org.sandboxpowered.silica.registry.RegistryEntry
import org.sandboxpowered.silica.state.property.Property

class SilicaStateFactory<B : RegistryEntry<B>, S : PropertyContainer<S>>(
    override val baseObject: B,
    map: Map<String, Property<*>>,
    stateCreator: Factory<B, S>
) : StateProvider<B, S> {
    private val propertiesByName: ImmutableSortedMap<String, Property<*>> = ImmutableSortedMap.copyOf(map)
    override val validStates: ImmutableList<S>
    override val baseState: S
        get() = this.validStates[0]

    interface Factory<B, S : PropertyContainer<S>> {
        fun create(base: B, values: ImmutableMap<Property<*>, Comparable<*>>): S

        companion object {
            fun <B, S : PropertyContainer<S>> of(factory: (base: B, properties: ImmutableMap<Property<*>, Comparable<*>>) -> S): Factory<B, S> {
                return object : Factory<B, S> {
                    override fun create(base: B, values: ImmutableMap<Property<*>, Comparable<*>>): S {
                        return factory.invoke(base, values)
                    }
                }
            }
        }
    }

    init {
        // Collect all possible values for set properties
        var sequence: Sequence<List<Pair<Property<*>, Comparable<*>>>> = sequenceOf(emptyList())
        propertiesByName.forEach { (_, property) ->
            sequence = sequence.flatMap { list ->
                property.values.map { comparable ->
                    list + (property to comparable)
                }
            }
        }

        val propertyToState: MutableMap<Map<Property<*>, Comparable<*>>, S> = LinkedHashMap()
        val allStates: MutableList<S> = ArrayList()

        // Generate combined states with all possible values
        sequence.forEach { list ->
            val propertyValues = list.stream()
                .collect(ImmutableMap.toImmutableMap(
                    { obj: Pair<Property<*>, Comparable<*>> -> obj.first },
                    { obj: Pair<Property<*>, Comparable<*>> -> obj.second }
                ))
            val state = stateCreator.create(baseObject, propertyValues)
            propertyToState[propertyValues] = state
            allStates.add(state)
        }

        // Initialize the tables within each state allowing them to know other state values
        allStates.forEach { state ->
            if (state is BaseState<*, *>) {
                (state as BaseState<B, S>).initTable(propertyToState)
            }
        }
        this.validStates = ImmutableList.copyOf(allStates)
    }
}